src/tools/cephfs_mirror/Mirror.cc

```diff
@@ -497,47 +496,45 @@ void Mirror::update_fs_mirrors() {
     ("cephfs_mirror_restart_mirror_on_blocklist_interval").count();
-  bool check_blocklist = blocklist_interval > 0 && ((now - m_last_blocklist_check) >= blocklist_interval);
-
   double failed_interval = g_ceph_context->_conf.get_val<std::chrono::seconds>
     ("cephfs_mirror_restart_mirror_on_failure_interval").count();
-  bool check_failure = failed_interval > 0 && ((now - m_last_failure_check) >= failed_interval);
 
   {
     std::scoped_lock locker(m_lock);
     for (auto &[filesystem, mirror_action] : m_mirror_actions) {
-      auto failed = mirror_action.fs_mirror && mirror_action.fs_mirror->is_failed();
-      auto blocklisted = mirror_action.fs_mirror && mirror_action.fs_mirror->is_blocklisted();
-
-      if (check_failure && !mirror_action.action_in_progress && failed) {
-        // about to restart failed mirror instance -- nothing
-        // should interfere
-        dout(5) << ": filesystem=" << filesystem << " failed mirroring -- restarting" << dendl;
-        auto peers = mirror_action.fs_mirror->get_peers();
-        auto ctx =  new C_RestartMirroring(this, filesystem, mirror_action.pool_id, peers);
-        ctx->complete(0);
-      } else if (check_blocklist && !mirror_action.action_in_progress && blocklisted) {
-        // about to restart blocklisted mirror instance -- nothing
-        // should interfere
-        dout(5) << ": filesystem=" << filesystem << " is blocklisted -- restarting" << dendl;
-        auto peers = mirror_action.fs_mirror->get_peers();
-        auto ctx = new C_RestartMirroring(this, filesystem, mirror_action.pool_id, peers);
-        ctx->complete(0);
+      auto failed_restart = mirror_action.fs_mirror && mirror_action.fs_mirror->is_failed() &&
+	(failed_interval > 0 && (mirror_action.fs_mirror->get_failed_ts() - now) > failed_interval);
+      auto blocklisted_restart = mirror_action.fs_mirror && mirror_action.fs_mirror->is_blocklisted() &&
+	(blocklist_interval > 0 && (mirror_action.fs_mirror->get_blocklisted_ts() - now) > blocklist_interval);
+
+      if (!mirror_action.action_in_progress && !_is_restarting(filesystem)) {
+	if (failed_restart) {
+	  // about to restart failed mirror instance -- nothing
+	  // should interfere
+	  dout(5) << ": filesystem=" << filesystem << " failed mirroring -- restarting" << dendl;
+	  _set_restarting(filesystem);
+	  auto peers = mirror_action.fs_mirror->get_peers();
+	  auto ctx =  new C_RestartMirroring(this, filesystem, mirror_action.pool_id, peers);
+	  ctx->complete(0);
+	}  else if (blocklisted_restart) {
+	  // about to restart blocklisted mirror instance -- nothing
+	  // should interfere
+	  _set_restarting(filesystem);
+	  dout(5) << ": filesystem=" << filesystem << " is blocklisted -- restarting" << dendl;
+	  auto peers = mirror_action.fs_mirror->get_peers();
+	  auto ctx = new C_RestartMirroring(this, filesystem, mirror_action.pool_id, peers);
+	  ctx->complete(0);
+
+	}
```

Comment:
@vshankar looks damn neat now but I have a suggestion here, the context completion code is repetitive in both if{} and else if{}, how about 

```suggestion
      if (!mirror_action.action_in_progress && !_is_restarting(filesystem)) {
	if (failed_restart) {
	  // about to restart failed mirror instance -- nothing
	  // should interfere
	  dout(5) << ": filesystem=" << filesystem << " failed mirroring -- restarting" << dendl;
	}  else if (blocklisted_restart) {
	  // about to restart blocklisted mirror instance -- nothing
	  // should interfere
	  dout(5) << ": filesystem=" << filesystem << " is blocklisted -- restarting" << dendl;
	}
	_set_restarting(filesystem);
	auto peers = mirror_action.fs_mirror->get_peers();
	auto ctx =  new C_RestartMirroring(this, filesystem, mirror_action.pool_id, peers);
	ctx->complete(0);
}
```

Reply:
fixed and updated a tad better change ;)

Reply:
much better than my version of better :D

Reply:
one question; if it case the daemon has failed(or blocklisted), dont you think the other info(blocklisted:false or failed:false) would be redundant?

Reply:
> one question; if it case the daemon has failed(or blocklisted), dont you think the other info(blocklisted:false or failed:false) would be redundant?

Not really, I deliberately kept it with dout(5). Maybe dout(10) would have been fine. Doesn't really matter IMO.

Reply:
what i meant was adding dout inside the `if (failed_restart || blocklisted_restart) {}`, then have something like 

```
dout(5) << "filesystem=" << filesystem << " mirroring failed due to mirror daemon << (failed ? "failed" : "blocklisted") << dendl;
```